### DISCUSSION TITLE
Alternation of Generations Model - trouble with subpopulations

### DISCUSSION THREAD

**Discussion by elissasoroj** [tskit-dev/pyslim#192] [Q&A] [answered]
Hi again, 

I am using an alternation of generations model in SLiM, which passes individuals between two subpopulations (p0 and p1 in my case). In generation 1 diploid sporophytes live in p1 and produce haploid gametophytes directly into p0, then the sporophytes die. In generation 2 the gametophytes then reproduce and place new diploid sporophytes back into p1 then the gametophytes die.  

This seems to cause problems at the recapitation step in pyslim, because half the nodes are assigned to Population 0 and half are assigned to Population 1, but they really should all be treated as the same population by msprime. 
I was not able to alter the population data in the tables, as I I got this error: `raise FrozenInstanceError()`.
I also had no luck setting a migration matrix of [[0, 1], [1, 0] or other migration rates of < 1 for recapitation. 

Any suggestions for how to proceed would be greatly appreciated.

Thanks!
~Elissa

**Comment by petrelharp** [comment, marked as answer]
Good question! Let's see - I imagine you're doing something like is described here:
  https://tskit.dev/pyslim/docs/latest/tutorial.html#recapitation-with-more-than-one-population
... is that right? I'd like to see what code you're running to get those errors. And, it sounds like what you'd actually like to do is to merge the two populations into one for the recapitation step, which we could do with an msprime `population_split` - I'd be happy to suggest how to do this, but it'd help to see what you're doing, first?

**Reply by elissasoroj** [reply]
Hi Peter,

Yes, the workflow I am using is similar to the documentation you linked to, but indeed I am trying to merge the populations into one for recapitation. I have copied my code below - 

This is what I get when I run recapitation on the treesequence from SLiM with a migration matrix of [[0,1.0], [1.0,0]]:

```python
# read in the sequence
ts = pyslim.load("../notebooks/bryo_branch2.trees")

#set initial population sizes to initial sizes of 1000 
pop_configs = [msprime.PopulationConfiguration(initial_size=1000)
    for _ in range(self.merged_ts.num_populations)]

matrix = [[0,1], [1,0]]

bryo_oldbranch_rts = bryo_oldbranch.recapitate(
    population_configurations=pop_configs,
    migration_matrix= matrix,
    recombination_rate=1e-9,
    random_seed=4,
)
```
This is what the resulting ts looks like after recapitation:
<img width="874" alt="Screen Shot 2021-08-06 at 12 57 26 PM" src="https://user-images.githubusercontent.com/71294603/128546039-c551e97e-8cba-455b-8dd8-5061f6cafe9c.png">

The result is the same for mutation rate of slightly less than 1.0, such as 0.9.

-----

I also tried to manually set the population at the nodes so that they are all the same by altering the tables:
```python
tables = ts.tables
for node in tables.nodes:
    node.population = 0
mod_ts = pyslim.load_tables(tables, slim_format=True)
for node in mod_ts.nodes():
    print(node.population)
```
And I get the following error:
```
---------------------------------------------------------------------------
FrozenInstanceError                       Traceback (most recent call last)
<ipython-input-107-c37311daa7df> in <module>
      1 tables = bryo_oldbranch.tables
      2 for node in tables.nodes:
----> 3     node.population = 0
      4 mod_ts = pyslim.load_tables(tables, slim_format=True)
      5 for node in mod_ts.nodes():

~/miniconda3/lib/python3.8/site-packages/attr/_make.py in _frozen_setattrs(self, name, value)
    551         Attached to frozen classes as __setattr__.
    552         """
--> 553         raise FrozenInstanceError()
    554 
    555 

FrozenInstanceError: 
```
I'm afraid the SLiM code is a beast, but please let me know if you would like me to share it for reference.

**Reply by petrelharp** [reply]
Ah, ok, makes sense!  Let's see: I can't tell what's wrong with the trees pictured in the first solution -- but maybe we can skip that, since it seems like that the second solution is more correct? The problem is that (unintuitively) the "node table row" objects returned by iterating over `table.nodes` are not modifiable, they are just "views" into the underlying data and you can't use them to modify the tables (yet). So, you'd need to do this:
```
tables = ts.tables
tables.nodes.population = np.zeros(tables.nodes.num_rows)
mod_ts = pyslim.load_tables(tables, slim_format=True)
for node in mod_ts.nodes():
    print(node.population)
``` 
Let me know if that fixes things?

**Reply by elissasoroj** [reply]
Thank you very much for the prompt reply. 

That code seemed to work (I had to remove `slim_format = True`, which doesn't seem to be a parameter of `load_tables()` anymore). After simplification I had one population in my ts, as I was hoping for!

I am still getting a similar result after recapitation, however. Perhaps I am just not used to looking at tree sequences and don't know what to expect. Our concern is this portion of the ts where no coalescence seems to happen on most of the branches. It seems like too many of the edges are of a similar age. Is this normal? 

Area of concern highlighted in blue:
![Screen Shot 2021-08-06 at 5 51 39 PM](https://user-images.githubusercontent.com/71294603/128575244-7bcb6835-2ced-4aff-8eb3-5db04b9b58b1.jpg)

And a smaller sample to make things clearer. That same section has no coalescence for many nodes, it seems (only one node show here):
![Screen Shot 2021-08-06 at 5 57 22 PM](https://user-images.githubusercontent.com/71294603/128575251-fa26bbb8-544a-40e8-815d-4c432cf4a8e4.jpg)

**Reply by petrelharp** [reply]
Well, my first interpretation was that maybe the short subtree was a small subpop in SLiM, the uncoalesced nodes were in a big subpop in SLiM, and recaptiation started at the top of the blue bit. But, thinking more about your particular model (since "populations" are actually life stages), I agree, this seems strange - I wonder if maybe you've got a bug in the model, so that individuals can get stuck in one life stage and never die? You could check this by looking at their `age` (e.g., `[ind.metadata['age'] for ind in ts.individuals_alive_at(0)]`).  It would also help to see where recapitation starts on the time axis of the figure.

**Reply by elissasoroj** [reply]
All the ages are '0' at time = 0 and none are alive at time = 1, which tracks with the model (all of them only live a single generation). I doubt that individuals are getting stuck in one life stage because we have `return F` in a `survival()` callback. 
Here is the tree with a time axis, and the start of the SLiM simulation marked in blue:
<img width="569" alt="Screen Shot 2021-08-06 at 10 25 47 PM" src="https://user-images.githubusercontent.com/71294603/128585040-eb2eaf34-e7a5-459e-a5d7-44defd228326.png">

Hmm, very interesting, the gap is happening in the SLiM part of the simulation. Perhaps because I neglected to have a burn-in period? 

Additionally, it seems like most of the mutations are occurring on a single branch. This is from a different sample than the screen-capture above:
<img width="768" alt="Screen Shot 2021-08-06 at 10 51 18 PM" src="https://user-images.githubusercontent.com/71294603/128585644-c67e9879-276c-40aa-b6dd-39a79b1d39f6.png">

I checked all the nodes from both samples and mutations only occur on that long branch farthest to the left and never on the other ones

---
Here is the SLiM code, just in case -
(apologies for length and note, the genomic elements in this script are different because the ones used in the code that generated the ts above are very long indeed; additionally, this uses the github build of SLiM)
```c
initialize() {
	initializeSLiMModelType("nonWF");
	
	initializeRecombinationRate(1e-09, 1000000);
	initializeMutationRate(1e-07);
	initializeTreeSeq();
	
	initializeMutationType('m6', 0.5, 'f', 0.0);
	m6.convertToSubstitution = T;
	initializeMutationType('m4', 0.2, 'n', -0.05, 0.015);
	m4.convertToSubstitution = T;
	
	initializeGenomicElementType('g11', m6, 1);
	initializeGenomicElementType('g10', m4, 1);
	
	types = rep(g10, asInteger(floor(3/3)));
	starts = 99997 + seqLen(integerDiv(3, 3)) * 3;
	ends = starts + 1;
	initializeGenomicElement(types, starts, ends);
	
	defineConstant('dK', 1000);
	defineConstant('hK', 1000);
	defineConstant('Death_chance', 0);
	defineConstant('FtoM', 0.5);
	defineConstant('Spore_num', 100);
	defineConstant('Clone_rate', 0.0);
	defineConstant('Self_rate', 0.0);
	defineConstant('Maternal_weight', 0);
}

// define Bryophyte subpops: diploid sporophytes, haploid gametophytes
1 early() { // executes after offspring are generated
	sim.addSubpop('p1', dK);
	sim.addSubpop('p0', hK);}

reproduction(p1 ) { // generates gametes from sporophytes
	g_1 = genome1;
	g_2 = genome2;
	meiosis_reps = floor(Spore_num/2);
	for (rep in 1:meiosis_reps)
	{
		breaks = sim.chromosome.drawBreakpoints(individual);
		p0.addRecombinant(g_1, g_2, breaks, NULL, NULL, NULL).tag = ifelse (runif(1)<FtoM, 1, 0);
		p0.addRecombinant(g_2, g_1, breaks, NULL, NULL, NULL).tag = ifelse (runif(1)<FtoM, 1, 0);
	}
}

reproduction(p0 ) { // generates gametes from sporophytes
	if (individual.tag == 1)    // females find male gametes to reproduce
	{
		reproduction_opportunity_count = 1;
		
		// clones give the focal individual extra opportunities to reproduce
		if (runif(1) <= Clone_rate)
			reproduction_opportunity_count = reproduction_opportunity_count + 1;
		
		for (repro in seqLen(reproduction_opportunity_count))
		{
			if (runif(1) <= Self_rate)
			{
				// this is selfing using two identical gametes – intragametophytic selfing
				// intergametophytic selfing might happen below, by chance
				p1.addRecombinant(individual.genome1, NULL, NULL, individual.genome1, NULL, NULL);
			}
			else
			{
				sperm = p0.sampleIndividuals(1, tag=0); // find a male!
				
				if (sperm.size() == 1)
				{
					child = p1.addRecombinant(individual.genome1, NULL, NULL, sperm.genome1, NULL, NULL);
					
					if (Maternal_weight > 0) //Mother's fitness affects sporophyte fitness; see survival()
						child.setValue("maternal_fitness", subpop.cachedFitness(individual.index));
					
					sperm.tag = 2;  // take out of the mating pool
				}
			}
		}
	}
}

early() {  // alternation of generations
	if (sim.generation % 2 == 0) //diploids (p1) just generated haploid gametophytes
	{
		//fitness affects gametophyte survival
		p0.fitnessScaling = (hK / p0.individualCount);
		
		//p0 and p1 survival callbacks
		s1.active = 1;
		s2.active = 0;
		s3.active = 1;
		s4.active = 0;
		
		// haploids get modified fitness, without dominance
		s5.active = 1;
		s6.active = 1;
	}
	else //odd generations = gametophytes (p0) just generated sporophytes
	{
		p1.fitnessScaling = dK / p1.individualCount; //fitness affects sporophytes
		
		//turn off p0 survival callbacks
		//turn on p1 survival callbacks
		s1.active = 0;
		s2.active = 1;
		s3.active = 0;
		s4.active = 1;
		
		// diploids get SLiM's standard fitness calculation, with dominance
		s5.active = 0;
		s6.active = 0;
	}
}

s1 survival(p1) { return F; }
s2 survival(p1){//this code implements random death chance
	if (runif(1) < Death_chance)
		return F;
	else
		return NULL;
	
	// maternal effect as weighted average
	maternal_effect = individual.getValue("maternal_fitness");
	
	if (!isNULL(maternal_effect))
	{
		corrected_fitness = (maternal_effect * Maternal_weight) + fitness * (1 - Maternal_weight);
		return (draw < corrected_fitness);
	}
	return NULL;
}

s3 survival(p0) //even
{//this code implements random death chance
	if (runif(1) < Death_chance)
		return F;
	else
		return NULL;
}

s4 survival(p0)  { return F; } //odd
s5 fitness(m6) { // gametophytes have no dominance effects
	return 1 + mut.selectionCoeff;}
s6 fitness(m4) { // gametophytes have no dominance effects
	return 1 + mut.selectionCoeff;}

late() { // fixes mutations in haploid gen
	if (sim.generation % 2 == 0) //gametophytes have just undergone fitness selection
	{
		muts5 = sim.mutationsOfType(m6);
		freqs5 = sim.mutationFrequencies(NULL, muts5);
		if (any(freqs5 == 0.5))
			sim.subpopulations.genomes.removeMutations(muts5[freqs5 == 0.5], T);
		muts6 = sim.mutationsOfType(m4);
		freqs6 = sim.mutationFrequencies(NULL, muts6);
		if (any(freqs6 == 0.5))
			sim.subpopulations.genomes.removeMutations(muts6[freqs6 == 0.5], T);
	}
}

10000 late() {  // end of sim; save individuals, save .trees file
	sim.treeSeqRememberIndividuals(sim.subpopulations.individuals);
	sim.treeSeqOutput('bryo_ancestralneutral.trees');}
```

**Reply by petrelharp** [reply]
Very nice model! I agree, definitely no-one is surviving more than one time step, and after thinking through the reproduction callbacks I don't see anything wrong. Now I wonder if you're seeing a selective sweep? I agree that the tree you sent this time looks a bit odd, but when I run this SLiM script and look at the resulting trees I don't see anything like what you're seeing. I'd need a reproducible example (like, exactly the same code that you ran to produce the figure you're seeing).

A small note - the second half of the `s2` block (with maternal fitness effects) never runs, if I'm reading it correctly. Tangentially, instead of returning F from the alternating survival callbacks, you could set their fitnessScalings to 0 - e.g., line 67 could have
```
		p0.fitnessScaling = (hK / p0.individualCount);
		p1.fitnessScaling = 0;
```

**Reply by elissasoroj** [reply]
Thanks! I’ve attached a SLiM script with a proper chromosome (sorry I didn’t provide earlier, it’s just so unwieldy). It generated this ts (full and then sample of 20 ends):

![Screen Shot 2021-08-07 at 3 42 50 PM](https://user-images.githubusercontent.com/71294603/128613186-5a41707a-19a5-4413-ac69-9bc22b7cfd3d.png)
<img width="481" alt="Screen Shot 2021-08-07 at 4 23 28 PM" src="https://user-images.githubusercontent.com/71294603/128613189-0111884c-b940-4cc7-949e-c786fcce5c21.png">

I guess it could be a selective sweep but it seems too consistent between runs. 

The scripts are kind of weird because they are generated in Python. If you’re wondering what the heck is going on with the genomic elements - I’ve written it so every third base in a coding region is empty, so we are simulating a proxy of synonymous and non-synonymous sites. And you’re right, although maternal fitness effects and random death are turned off in this particular model (because the constants are set to 0), they are still in there because they are options in the Python program that generated the script. Additionally, this is one of several different default models we have that simulate plant life cycles. The Python program that generates the SLiM script can use any of the default models, so although it would definitely be more efficient to use fitnessScaling, it is separated out in a survival() callback because we need to toggle it on and off for different models.

[link to SLiM script](https://github.com/elissasoroj/shadie/blob/main/example_scripts/bryophyte_model_nonneutral.slim)

*sorry for the double post

**Reply by petrelharp** [reply]
Ah-ha: notice that all the uncoalesced lineages are even, and the ones that coalesce before recapitation are odd. Loading in the output from your slim script shows the same thing except the other way around (even ones coalesce):
```
ts = pyslim.load("bryo_branch4.trees")
t = ts.first()
for i in range(10):
    for n in ts.individual(i).nodes:
        print(n, t.parent(n))
```
gives
```
0 2492
1 -1
2 2184
3 -1
4 2184
5 -1
6 2511
7 -1
8 2813
9 -1
# etcetera
```
Since adjacent nodes are the two chromosomes of a single individual, it looks to me like your reproduction callbacks are making it so that the second chromosomes of each individual never coalesce. And, looking into things - it looks like this has something to do with `addRecombinant` for sporophytes adding in "empty" genomes. At first I thought that those weird branches are just the "empty" genomes, and so they'd not appear if I wrote out the tree sequence at an odd time instead of an even time, but that did not fix the problem.

So, I'm not quite sure where the problem is. If the problem is that you just need to ignore the second chromosomes of each individual, then `simplify( )` will take care of that for you (but my quick try at that didn't fix the problem). Happily you don't need your big complicated script to debug this - you can make a very small example with like 5 individuals that runs for 4 generations (or something); if it were me I would (a) ask SLiM to please print out all parent-child relationships; and (b) `remember` everyone in every generation, so then I could compare what's in the tree sequence to what SLiM prints out to understand what's going on.

Let me know if the problem isn't clear.

**Reply by petrelharp** [reply]
Ah-ha, again: setting the population sizes to 10 and replacing the last blocks in your slim script with:
```
6 late() { // executes after selection occurs
	sim.treeSeqRememberIndividuals(sim.subpopulations.individuals)
		;
	sim.treeSeqOutput('test6.trees');
}
7 late() { // executes after selection occurs
	sim.treeSeqRememberIndividuals(sim.subpopulations.individuals)
		;
	sim.treeSeqOutput('test7.trees');
}
```
and then doing:
```
ts7 = pyslim.load("test7.trees")
ts6 = pyslim.load("test6.trees")
for ts in (ts6, ts7):
  for i in ts.individuals_alive_at(0):
    ind = ts.individual(i)
    print(ind.id, ind.metadata['subpopulation'])
  print("======")
```
gets me
```
0 0
1 0
2 0
3 0
4 0
=====
0 1
1 1
=====

```
which is telling me that in even time steps, you have only "haploid" gametes alive; so your weird branches are from those "empty" genomes; while in odd time steps you have diploids. The easy solution is to just output the tree sequence at an odd time step.

**Reply by elissasoroj** [reply]
Ah, of course! I had forgotten about the empty genomes in the haploids. Thank you very much for all your help - with the original question and all my follow up questions!
